lms/djangoapps/courseware/tabs.py

```diff
@@ -3,44 +3,347 @@
 perform some LMS-specific tab display gymnastics for the Entrance Exams feature
 """
 from django.conf import settings
-from django.test.client import RequestFactory
 from django.utils.translation import ugettext as _
 
 from courseware.access import has_access
 from courseware.entrance_exams import user_must_complete_entrance_exam
-from student.models import CourseEnrollment, EntranceExamConfiguration
-from xmodule.tabs import CourseTabList
+from openedx.core.djangoapps.course_views.course_views import CourseViewTypeManager, CourseViewType, StaticTab
+from student.models import CourseEnrollment
+from xmodule.tabs import CourseTab, CourseTabList, key_checker
 
-from util import milestones_helpers
+
+class EnrolledCourseViewType(CourseViewType):
+    """
+    A base class for any view types that require a user to be enrolled.
+    """
+    @classmethod
+    def is_enabled(cls, course, user=None):
+        if not user:
+            return True
+        return CourseEnrollment.is_enrolled(user, course.id) or has_access(user, 'staff', course, course.id)
+
+
+class CoursewareViewType(EnrolledCourseViewType):
```

Comment:
Are you asking in general, or about "Courseware" in particular? For the general name, I wanted to generalize the concept to not be about tabs, but to instead be about views that can be added to the course. My intention is that these views do not necessarily have to be shown as tabs, particularly on mobile where there may be a completely different UX metaphor. I settled on calling the entry point ``CourseViewType`` for that reason.

For "Courseware" in particular, it seemed reasonable to name the subclass to match its superclass. Note that this isn't the tab itself (that is an instance of ``CourseViewTab``), but is instead information about the view that can be shown as a tab.

Does that help? Would you prefer a different name?